postgresql/postgresql_plugin.cc

```diff
@@ -0,0 +1,832 @@
+        return CellValue(atoi(PQgetvalue(res, i, j)));
+    }
+    else if (postgrestype == 701) {
+        return CellValue(atof(PQgetvalue(res, i, j)));
+    }
+    else {
+        return CellValue(PQgetvalue(res, i, j));
+    }
+}
+
+pg_conn* startPostgresqlConnection(string databaseName, string host, int port)
+{
+    auto creds = getCredential("postgresql", host);
+
+    string connString;
+    connString += "dbname=" + databaseName
+               + " host=" + host
+               + " port=" + std::to_string(port)
+               + " user=" + creds.id
+               + " password=" + creds.secret;
+
+    POSTGRESQL_VERBOSE(cerr << connString << endl;)
+
+    auto conn = PQconnectdb(connString.c_str());
+    if (PQstatus(conn) != CONNECTION_OK)
+    {
+        string errorMsg(PQerrorMessage(conn));
+        PQfinish(conn);
+        throw HttpReturnException(400, "Could not connect to Postgresql database: ", errorMsg);
+    }
+
+    POSTGRESQL_VERBOSE(cerr << "connection successful!" << endl;)
+    return conn;
+}
+
+}
+
+/*****************************************************************************/
+/* POSTGRESQL DATASET                                                        */
+/*****************************************************************************/
+
+struct PostgresqlDatasetConfig {
+    string databaseName;
+    int port;
+    string host;
+    string tableName;
+    string primaryKey;
+
+    PostgresqlDatasetConfig() {
+        databaseName = "database";
+        port = 1234;
+        host = "localhost";
+        tableName = "mytable";
+        primaryKey = "key";
+    }
+};
+
+DECLARE_STRUCTURE_DESCRIPTION(PostgresqlDatasetConfig);
+DEFINE_STRUCTURE_DESCRIPTION(PostgresqlDatasetConfig);
+
+PostgresqlDatasetConfigDescription::
+PostgresqlDatasetConfigDescription()
+{
+    addField("databaseName", &PostgresqlDatasetConfig::databaseName, "Name of the database to connect to.");
+    addField("port", &PostgresqlDatasetConfig::port, "Port of the database to connect to.", 1234);
+    addField("host", &PostgresqlDatasetConfig::host, "Address of the database to connect to ");
+    addField("tableName", &PostgresqlDatasetConfig::tableName, "Name of the table to be recorded into");
+    addField("primaryKey", &PostgresqlDatasetConfig::primaryKey, "Primary key used to access progress table");
+}
+
+struct PostgresqlDataset: public Dataset {
+
+    PostgresqlDatasetConfig config_;
+
+    PostgresqlDataset(MldbServer * owner,
+                 PolyConfig config,
+                 const std::function<bool (const Json::Value &)> & onProgress)
+        : Dataset(owner)
+    {
+        config_ = config.params.convert<PostgresqlDatasetConfig>();
+    }
+    
+    virtual ~PostgresqlDataset()
+    {
+    }
+
+    pg_conn* startConnection() const
+    {
+        return startPostgresqlConnection(config_.databaseName, config_.host, config_.port);
+    }
+
+    virtual Any getStatus() const
+    {
+        return string("ok");
```

Comment:
I would create an MLDBFB ticket to return a better status.  I feel this is particularly important here since many things can go wrong.